### ISSUE TITLE
rfc: automating the switching of inputs to the device of the params

### ISSUE THREAD

**Issue by stas00** [pytorch/pytorch#49961]
## 🚀 Feature

tldr: figure out a way for pytorch to automatically switch arguments of pytorch operations to the device of the params involved in the operation. Primarily to ease the facilitation of model parallelism, but may have other uses.

The full story:

1. We have the simple case of having a model run on a single GPU (DDP-included) and needing to switch model inputs to that GPU. a single `.to(device)` call and we are done.

2. Then we have model parallelism where different layers will be scattered across multiple devices. Things become very complicated now because it's far from trivial to know which device the inputs need to be switched to.

The most simple approach that will always work is to just always add: `to(self.get_device)` to all input arguments in `forward`, e.g.:

```
def self.get_device(self): return next(self.parameters(recurse=True)).device
def forward(self, input):
    x = self.linear(input.to(self.get_device))
```
this would work without needing `if MP: ....` with a slight overhead.

3. but this makes the code hard to read and so why not do this automatically before `forward` as in:

```
def model_parallel_call(self, *input, **kwargs):

    try:
        device = next(self.parameters(recurse=True)).device
    except StopIteration:
        device = None

    if device is not None:

        input = list(input)
        for i, v in enumerate(input):
            if v is not None:
                input[i] = v.to(device)
        input = tuple(input)

        for k in kwargs.keys():
            if kwargs[k] is not None and torch.is_tensor(kwargs[k]):
                kwargs[k] = kwargs[k].to(device)

    return nn.Module.__call__(self, *input, **kwargs)
    
@model_parallel_call
def forward(self, input):
    x = self.linear(x)    
```
pass as many inputs as you want and they are already on the right device MP or not.

So now we get an automatic switch of all inputs to the same device as the params of the model. We assume here that it is so - that is that all params of that layer are on the same model. 

But this is not always the case, as for example shown in the trivial example of https://pytorch.org/tutorials/intermediate/model_parallel_tutorial.html#apply-model-parallel-to-existing-modules

4. So my thinking is this - the next stage is to extend pytorch ops to automatically switch the inputs to the device of the params at the point of use, so if there is a `matmul` look at the device of the "designated" argument and switch all the other args to the same device, rather than assert that they have to be on the same device.

I totally understand that the devil is in the detail and how could one figure out what that "designated" argument, so that say we remain on gpu1 if the other arg is on cpu, how do we define DWIM (do what I mean) since it's not always the first arg that "leads". What I'm trying to communicate is that each op needs a way to check which argument sets the device for the rest of the arguments of that op.

Or perhaps there should be a way of switching some global flag that unless specified otherwise will switch all args to that device before doing any operation.

The goal of this is to be able to implement model parallelism by simply spreading out the params of a single model through multiple devices and having pytorch automagically sort out the rest.

I realize that perhaps my vision is not quite feasible at the moment, but perhaps there are some ways that can ease this process.

Thank you for reading and if you have some inspirations please don't hesitate to share.

**Comment by mruberry** [comment]
Hey @stas00, thanks for this RFC!

Implementing behavior like this would be tricky because a core tenet of PyTorch's UX is that it avoids implicit cross-device data movement. Sneakily moving tensors across devices would be a source of bugs, performance cliffs, and user confusion.

It would be nice to better support model parallelism, though, maybe with a more explicit mechanism like your code snippet demonstrates.

**Comment by stas00** [comment]
Thank you for the followup, @mruberry.

I certainly understand that such endeavour would be tricky and surely will have its portion of bugs to start with, but as the models keep on growing at a faster rate than gpu's memory it's very possible that model parallelism will become more of a norm than an oddity in the recent future. And as it can be seen very quickly from the [mp tutorial](https://pytorch.org/tutorials/intermediate/model_parallel_tutorial.html#apply-model-parallel-to-existing-modules) that kind of code looks already hairy in simple examples, and for complex models it quickly becomes a mess.

And surely such possible extension of implicit cross-device data movement (love the term!) support doesn't have to interfere with the normal solid pytorch and only to be enabled when desired.

Also if someone has experience and ideas wrt easier MP, we have been discussing this subject matter about the ongoing MP-implementation in `transformers` - t5 and gpt2 have already been ported to MP by @alexorona and we are trying to generalize the approach to other architectures. There are two threads at the moment:
* https://github.com/huggingface/transformers/issues/8771#issuecomment-752309731
* https://github.com/huggingface/transformers/pull/9323#issuecomment-752352280

**Comment by stas00** [comment]
And meanwhile I have much improved versions of the bring-data-to-the-right-device magical and explicit versions:
```
def recursive_to(device, item):
    """
    Switch any tensors found in `item` to `device`.
    Currently can handle a single tensor, or any of the nested list, tuple and dict structures.
    """

    if torch.is_tensor(item):
        return item.to(device)

    elif isinstance(item, list):
        for i, x in enumerate(item):
            item[i] = recursive_to(device, x)
        return item

    elif isinstance(item, tuple):
        return tuple(recursive_to(device, list(item)))

    elif isinstance(item, dict):
        for k, v in item.items():
            item[k] = recursive_to(device, v)
        return item

    else:
        return item


def model_parallel_inputs_to_device(func):
    """
    This decorator is a noop unless self.model_parallel == True.

    It will try to find a at least one parameter or a buffer to read layer's .device from and then will
    automatically copy any inputs to that device before `forward` is called.

    this will work do its magical thing only if all params of this layer are on the same device
    """

    def _call__mp(self, *input, **kwargs):

        if not hasattr(self, "model_parallel") or not self.model_parallel:
            return func(self, *input, **kwargs)

        # get device of any of the param of this layer
        try:
            device = next(self.parameters(recurse=True)).device
        except StopIteration:
            device = None

        # print(f"layer device: {device}")
        if device is not None:
            # torch.cuda.set_device(device)
            # print(f"auto-move inputs to {device}")

            input = recursive_to(device, input)
            kwargs = recursive_to(device, kwargs)

            return func(self, *input, **kwargs)

    return _call__mp


def model_parallel_inputs_to_specific_device(device, *input):
    #print(f"move specific inputs to {device}")
    output = recursive_to(device, input)
    # remove the need for the caller to perform "a, = foo(a)", which otherwise will make `a` a tuple when it might be not be
    return output[0] if len(output) == 1 else output


def model_parallel_call(self, *input, **kwargs):

    # get device of any of the param of this layer
    try:
        device = next(self.parameters(recurse=True)).device
    except StopIteration:
        device = None

    # print(f"layer device: {device}")
    if device is not None:
        # torch.cuda.set_device(device)
        input = recursive_to(device, input)
        kwargs = recursive_to(device, kwargs)

    return nn.Module.__call__(self, *input, **kwargs)


```

**Comment by stas00** [comment]
Also it looks like some pytorch and apex ops are buggy and one has to use the deprecated `torch.cuda.set_device` to overcome corrupt memory and illegal memory access failures when using multiple devices as moving all parts `.to()` device doesn't seem to be enough in those cases. e.g. see [this](https://github.com/NVIDIA/apex/issues/1022) and [this](https://github.com/pytorch/pytorch/issues/21819).

So if pytorch already has a global arbiter `torch.cuda.set_device`, albeit deprecated, why not just use it to set the current device and have all the ops move the data to that device before doing any operation (most of the time they won't need to do).

Perhaps that is probably too wild, but then perhaps there could be a context manager, so it can be pretty tightly controlled scope-wise. There is the `torch.cuda.device` context manager, but this one won't move data, so perhaps there could be a different context manager that will do this instead.

**Comment by mruberry** [comment]
Thank you for reporting the Apex issue, @stas00. I agree that something like a context manager would be an interesting UX to consider to address this. Let me add the "needs design" label to signal we should brainstorm some alternatives.

**Comment by ngimel** [comment]
#27878 is a related issue discussing context manager forcing created tensor to be on a specified device. Here, you want not only newly created tensor to be on the specified device, but also all existing tensors be moved to it. A context manager like this would shadow legitimate coding bugs where tensors are on the wrong device and author should know about this, it would also make life of library developers harder - they'd need to develop their libraries to work when tensors can unexpectedly switch devices under them. 
On a side note, illegal memory access in Apex is definitely a bug, however, the fact that pytorch native functions work in this situation just masks that your default device is not what it should be. For every function, you are switching to a device of the tensor argument, and then switching back. That's a performance penalty, even if functionally things work.

**Comment by osalpekar** [comment]
Just providing some insight into the Model Parallelism aspect here - PyTorch has some primitives that allow you to enable model parallel training. You can place different model shards (groups of layers) on specific devices using the [RPC framework](https://pytorch.org/docs/stable/rpc.html) and maintain references to each shard. The RPC framework can also be used to move activations between shards, and [Distributed Autograd](https://pytorch.org/docs/stable/rpc.html#distributed-autograd-framework) and [Distributed Optimizer](https://pytorch.org/docs/stable/rpc.html#module-torch.distributed.optim) can be used to run the backward pass and optimizer across a sharded model (with relatively minimal code changes compared to single-node training).

Placing model shards on intended devices will also get easier with the Remote Module API (currently private [here](https://github.com/pytorch/pytorch/blob/00651b8c93354da13ab30c3fefd8cd529dc883e7/torch/distributed/nn/api/remote_module.py#L33)). cc @lw @beauby about GPU support for RPC

**Comment by stas00** [comment]
> #27878 is a related issue discussing context manager forcing created tensor to be on a specified device. Here, you want not only newly created tensor to be on the specified device, but also all existing tensors be moved to it. A context manager like this would shadow legitimate coding bugs where tensors are on the wrong device and author should know about this, it would also make life of library developers harder - they'd need to develop their libraries to work when tensors can unexpectedly switch devices under them.

Thank you for that reflection and the link to the discussion about a different context manager.

Just to clarify, I'm in no way proposing to change any current behavior and request for changes that will would lead to a more difficult world. The current code works mostly just fine, albeit requiring a lot of code which I feel could be not required. And there more code one has to write the more opportunities for bugs there is. 

I'm asking if the future pytorch could support other concepts other than explicit setting of all the arguments to a specific device before they can be operated on.

> On a side note, illegal memory access in Apex is definitely a bug, however, the fact that pytorch native functions work in this situation just masks that your default device is not what it should be. 

Can pytorch assert when such situation happens, so no such bugs are masked? My guess is that would require each pytorch op to do such check and it'd probably be inefficient and ugly.

> For every function, you are switching to a device of the tensor argument, and then switching back. That's a performance penalty, even if functionally things work.

Are you referring to my bug reproduction script or the topic of this rfc? 

Surely the intention is not to copy the data back and forth, but only when needed. A context manager could take care of it.

A simplified explanation would be with the usual drawing of the deep nn:
```
blocks   | [blk] ... [blk 2] | [blk 3] ... [blk 5] | [blk 6] ... [blk 7] | [head]
devices  |         0         |          1          |          2          |   0
```

Then it depends on the what parts are being used in each block, if they are not shared, there is no need to copy ever, other than at the boundaries.

---------------------

You mentioned the default device several times. I don't think it's documented too well. In fact if you look at the first 3 lines of:
https://pytorch.org/docs/stable/notes/cuda.html#cuda-semantics

```
cuda = torch.device('cuda')     # Default CUDA device
cuda0 = torch.device('cuda:0')
cuda2 = torch.device('cuda:2')  # GPU 2 (these are 0-indexed)
```
it's very ambiguous and doesn't tell me anything. What is the default device in that example?

**Comment by ngimel** [comment]
>Are you referring to my bug reproduction script or the topic of this rfc?
I was referring to your bug reproduction script. Default device is the device you are setting with `torch.cuda.set_device()`. It's possible to set device to `1` and then operate on the tensors on device `0`, but for every function internally pytorch would be calling `cudaSetDevice(0)` - launch function kernel - `cudaSetDevice(1)` as part of setting device guards, and this is generally less efficient then setting device to `0` in the first place.

**Comment by stas00** [comment]
> Just providing some insight into the Model Parallelism aspect here - PyTorch has some primitives that allow you to enable model parallel training. You can place different model shards (groups of layers) on specific devices using the [RPC framework](https://pytorch.org/docs/stable/rpc.html) and maintain references to each shard. The RPC framework can also be used to move activations between shards, and [Distributed Autograd](https://pytorch.org/docs/stable/rpc.html#distributed-autograd-framework) and [Distributed Optimizer](https://pytorch.org/docs/stable/rpc.html#module-torch.distributed.optim) can be used to run the backward pass and optimizer across a sharded model (with relatively minimal code changes compared to single-node training).
> 
> Placing model shards on intended devices will also get easier with the Remote Module API (currently private [here](https://github.com/pytorch/pytorch/blob/00651b8c93354da13ab30c3fefd8cd529dc883e7/torch/distributed/nn/api/remote_module.py#L33)). cc @lw @beauby about GPU support for RPC

Thank you very much for this insight. I haven't studied this approach yet. And will certainly do so.

Currently I've been trying to generalize an implementation of MP that was initially implemented by switching devices in just one process. (HF `transformers`). It's not too bad, but a bit of a pain to port to that approach as one has to hunt down all parts and switch them to the device the layer/block is mapped to. And I haven't seen it myself, but @alexorona has done all the heavy lifting reports that he had to inject `torch.cuda.set_device(device)` in multiple places as things were crashing for him without that, which indicates either bugs in pytorch or elsewhere.

**Comment by stas00** [comment]
> > Are you referring to my bug reproduction script or the topic of this rfc?
> 
> I was referring to your bug reproduction script. Default device is the device you are setting with `torch.cuda.set_device()`. It's possible to set device to `1` and then operate on the tensors on device `0`, but for every function internally pytorch would be calling `cudaSetDevice(0)` - launch function kernel - `cudaSetDevice(1)` as part of setting device guards, and this is generally less efficient then setting device to `0` in the first place.

That's super-useful and needs be absolutely documented - thank you, @ngimel 

So in the [mp tutorial](https://pytorch.org/tutorials/intermediate/model_parallel_tutorial.html#apply-model-parallel-to-existing-modules), the code needs to be changed to call `torch.cuda.set_device()` before switching to using the new device?

And so you're saying that we should do the same everywhere when we switch a whole set of operations to a new device. That is it's not enough to move (1) the model and (2) the data `.to(device)`, we have to (3) explicitly change the default device too.

Please correct me if I'm wrong.

**Comment by pritamdamania87** [comment]
@stas00 Would love to understand some more details about the model parallelism use case here. We're working on making model parallelism much easier in PyTorch and understanding some details of your use case might help us building higher level APIs better. 

For example, we're adding pipeline parallelism to PyTorch: https://github.com/pytorch/pytorch/blob/master/torch/distributed/pipeline/sync/pipe.py#L138 (will be released in 1.8) and this API needs you to only place modules on different devices and the pipeline API under the hood takes care of moving the data across devices.

> Also if someone has experience and ideas wrt easier MP, we have been discussing this subject matter about the ongoing MP-implementation in transformers - t5 and gpt2 have already been ported to MP by @alexorona and we are trying to generalize the approach to other architectures. There are two threads at the moment:

> huggingface/transformers#8771 (comment)
> huggingface/transformers#9323 (comment)

Based on this comment, it does seem like pipeline parallelism might be a good fit here since it has been used to train transformer models. Could you take a look at the pipeline parallelism API and see if it fits your needs? If pipeline parallelism is not the right strategy for you, we'd love to understand what sort of model parallelism scheme would fit your needs best. For example, does your graph contain several parallel branches such that you can execute it in parallel across multiple devices without needing pipelining?

**Comment by stas00** [comment]
>  @stas00 Would love to understand some more details about the model parallelism use case here. 

Yes, of course, here is the usecase of the PR I have been working on

Here is an example of a `sshleifer/distilbart-xsum-6-6` `BartForConditionalGeneration` model:

```
 (model): BartModel(
    (shared): Embedding(50264, 1024, padding_idx=1)
    (encoder): BartEncoder(
      (embed_tokens): Embedding(50264, 1024, padding_idx=1)
      (embed_positions): BartLearnedPositionalEmbedding(1026, 1024, padding_idx=1)
      (layers): ModuleList( 6 x BartEncoderLayer)
      (layernorm_embedding): FusedLayerNorm(torch.Size([1024]), eps=1e-05, elementwise_affine=True)
    )
    (decoder): BartDecoder(
      (embed_tokens): Embedding(50264, 1024, padding_idx=1)
      (embed_positions): BartLearnedPositionalEmbedding(1026, 1024, padding_idx=1)
      (layers): ModuleList( 6 x BartDecoderLayer)
      (layernorm_embedding): FusedLayerNorm(torch.Size([1024]), eps=1e-05, elementwise_affine=True)
    )
  )
  (lm_head): Linear(in_features=1024, out_features=50264, bias=False)
)
```
Note that I collapsed the huge bulk of it and it's represented by just 2 lines:
```
      (layers): ModuleList( 6 x BartEncoderLayer)
      (layers): ModuleList( 6 x BartDecoderLayer)
```
this is some 90% of the model and that's what we want to spread out through multiple gpus.

If you want to see the whole huge thing, it's just:
```
from transformers import BartForConditionalGeneration
mname = "sshleifer/distilbart-xsum-6-6"
model = BartForConditionalGeneration.from_pretrained(mname)
print(model)
```

So we have the bulk of memory used by 6 x `BartEncoderLayer` and 6 x `BartDecoderLayer`, plus some other components.

For the simplicity of the example let's say we have 2 gpus we want to split the model into.

Currently the idea is to put the 6 encoder layers on gpu 0 and the same for decoder layers but on gpu 1:
```
device_map = {
        "encoder": {0: [0, 1, 2, 3, 4, 5] },
        "decoder": {1: [0, 1, 2, 3, 4, 5] },
    }
```
or alternatively, splice each group as following:
```
device_map = {
        "encoder": {0: [0, 1, 2],
                    1: [3, 4, 5],
                    },
        "decoder": {0: [0, 1, 2],
                    1: [3, 4, 5],
                    },
    }
```
and the remaining specific layers can be all on gpu 0 or grouped closer to where they are needed. We still haven't quite finalized that map.

Of course, other models may have more or less layers and they don't have to have the same number of layers in encoder and decoder.

So now after the model is initialized - we turn MP on by doing `.to()` to all the layers (and their sub-modules) according to the device map and then in `forward` we switch inputs to the same device as the layer's params using a handy wrapper I shared here: https://github.com/pytorch/pytorch/issues/49961#issuecomment-753441248
with a little bit of fiddling with some of the params which happen to be shared, this is all that needs to be done to make the model parallel. Now I understand from @ngimel's comments that I need to add torch.cuda.set_device() to make it efficient. I think I can hide those inside the wrapper that moves all inputs to the right device.

Here is a work-in-progress https://github.com/huggingface/transformers/pull/9384 Bart port - it works just fine, just needs a bit of cleanup...

Please let me know if any other details would be helpful.

and my fantasy is, of course, is that it should be enough to map the layers to the right devices and have pytorch figure out the `.to()` for inputs auto-magically.

**Comment by pritamdamania87** [comment]
@stas00 Thanks a lot for providing details of the model and how you plan to use model parallelism. I'm assuming the BartEncoderLayer/BartDecoderLayer are sequential and if so, wouldn't one GPU always be idle?

> device_map = {
        "encoder": {0: [0, 1, 2, 3, 4, 5] },
        "decoder": {1: [0, 1, 2, 3, 4, 5] },
    }

In this example, while the encoder is running GPU1 is idle.

> device_map = {
        "encoder": {0: [0, 1, 2],
                    1: [3, 4, 5],
                    },
        "decoder": {0: [0, 1, 2],
                    1: [3, 4, 5],
                    },
    }

In this example, GPU1 is idle when parts 0,1 and 2 of the encoder are running. 

As I mentioned in my previous comment, have you evaluated using pipeline parallelism? It would provide better utilization of GPUs and you don't have to worry about `.to()` either.

**Comment by stas00** [comment]
Hmm, you're correct, I guess that was a naive implementation that basically just extends the available gpu memory but isn't taking advantage of the extra hardware. We definitely could do better.

I've to finish a few other things and will experiment with PP next. 

We also have PP in `fairscale` and `deepspeed` which we are now being integrating into the HF trainer, perhaps we could use those if there will be an objection to requiring pt-1.8.

**Comment by pritamdamania87** [comment]
> @pritamdamania87, I trust I can figure out the new API, but if there are existing examples/tutorials that always helps to save time. If none are available it's no problem.

We don't have tutorials yet, but will have some as part of pt1.8. We do have a benchmark here that you can look at as an example in the meantime: https://github.com/pytorch/pytorch/blob/master/benchmarks/distributed/pipeline/pipe.py.

> We also have PP in fairscale and deepspeed which we are now being integrating into the HF trainer, perhaps we could use those if there will be an objection to requiring pt-1.8, since it might be easier to ask for a specific version of the 3rd-party module if someone wants to use MP/PP.

The PP in PyTorch basically pulls in the PP work from fairscale, although if you'd like to test things out in PyTorch you can always use the nightly builds to use the current PP implementation in master.

**Comment by stas00** [comment]
> We don't have tutorials yet, but will have some as part of pt1.8. We do have a benchmark here that you can look at as an example in the meantime: https://github.com/pytorch/pytorch/blob/master/benchmarks/distributed/pipeline/pipe.py.

This is very helpful - thank you, @pritamdamania87!

> The PP in PyTorch basically pulls in the PP work from fairscale, although if you'd like to test things out in PyTorch you can always use the nightly builds to use the current PP implementation in master.

Oh, great, I was actually thinking of experimenting with fairscale's PP first, since we have just integrated their Sharded DDP into the HF trainer. So if we start using PP it might be fairscale if pt>=1.8 restriction won't be accepted by the HF team, since it'd be easier to ask to install a 3rd party module than request users to upgrade to pt-1.8 - I could be wrong.

**Comment by xuxy09** [comment]
> > Are you referring to my bug reproduction script or the topic of this rfc?
> 
> I was referring to your bug reproduction script. Default device is the device you are setting with `torch.cuda.set_device()`. It's possible to set device to `1` and then operate on the tensors on device `0`, but for every function internally pytorch would be calling `cudaSetDevice(0)` - launch function kernel - `cudaSetDevice(1)` as part of setting device guards, and this is generally less efficient then setting device to `0` in the first place.

Hi, can you explain this in more details? I am quite confused. If my default device is 1, and my model and data are both on device 0, could you better explain the internal process of Pytorch if I run the model? Do you mean the computation will happen on device 1 instead of 0?
